src/log_rate.c

```diff
@@ -0,0 +1,110 @@
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <unistd.h>
+#include <string.h>
+#include <time.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include "utils.h"
+#include "log_rate.h"
+#include "ctr_logging.h"
+
+#define IO_BUF_SIZE      65536
+#define BYTES_PER_PERIOD 1024
+#define SECS_PER_PERIOD  1
+#define BILLION          1000000000
+
+static const struct timespec secs_per_period = {SECS_PER_PERIOD, 0};
+static size_t bytes_written_this_period = 0;
+static struct timespec start_of_this_period;
+
+static struct timespec subtract_timespecs(const struct timespec* first, const struct timespec* second);
+static int64_t subtract_timespecs_nano(const struct timespec* first, const struct timespec* second);
+static void write_io_bufs(stdpipe_t pipe, char* buf, ssize_t count);
+static void sleep_for_the_rest_of_this_period();
+static void start_new_period();
+
+void log_rate_init() {
+	start_new_period();
+}
+
+bool log_rate_write_to_logs(stdpipe_t pipe, char *buf, ssize_t num_read) {
+	char* buf_start = buf;
+	ssize_t bytes_remaining = num_read;
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	int64_t diff_nano = subtract_timespecs_nano(&now, &start_of_this_period);
+	if (diff_nano < SECS_PER_PERIOD * BILLION) {
+		ssize_t bytes_we_can_write = BYTES_PER_PERIOD - bytes_written_this_period;
+		if (num_read <= bytes_we_can_write) {
+			write_io_bufs(pipe, buf_start, num_read);
```

Comment:
Does this assume that after a write we are still in the same period?

It seems that we have to re-evaluate what period we are in following each `write_to_logs()` call.